tagmanager/ctags/powershell.c

```diff
@@ -0,0 +1,660 @@
+/*
+*   Copyright (c) 2015, Enrico Tröger <enrico.troeger@uvena.de>
+*
+*   Loosely based on the PHP tags parser since the syntax is somewhat similar
+*   regarding variable and function definitions.
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License.
+*
+*   This module contains code for generating tags for Windows PowerShell scripts.
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#include "general.h"  /* must always come first */
+#include "main.h"
+#include "parse.h"
+#include "read.h"
+#include "vstring.h"
+#include "keyword.h"
+#include "entry.h"
+#include <string.h>
+
+#define SCOPE_SEPARATOR ":"
+
+
+#define ARRAY_LENGTH(array) (sizeof array / sizeof array[0])
+
+#define ACCESS_UNDEFINED NULL
+static const char *const accessTypes[] = {
+	ACCESS_UNDEFINED,
+	"global",
+	"local",
+	"script",
+	"private"
+};
+
+typedef enum {
+	K_FUNCTION,
+	K_VARIABLE,
+	COUNT_KIND
+} powerShellKind;
+
+static kindOption PowerShellKinds[COUNT_KIND] = {
+	{ TRUE, 'f', "function",	"functions" },
+	{ TRUE, 'v', "variable",	"variables" }
+};
+
+
+typedef enum eTokenType {
+	TOKEN_UNDEFINED,
+	TOKEN_EOF,
+	TOKEN_CHARACTER,
+	TOKEN_CLOSE_PAREN,
+	TOKEN_SEMICOLON,
+	TOKEN_COLON,
+	TOKEN_COMMA,
+	TOKEN_KEYWORD,
+	TOKEN_OPEN_PAREN,
+	TOKEN_OPERATOR,
+	TOKEN_IDENTIFIER,
+	TOKEN_STRING,
+	TOKEN_PERIOD,
+	TOKEN_OPEN_CURLY,
+	TOKEN_CLOSE_CURLY,
+	TOKEN_EQUAL_SIGN,
+	TOKEN_OPEN_SQUARE,
+	TOKEN_CLOSE_SQUARE,
+	TOKEN_VARIABLE
+} tokenType;
+
+typedef struct {
+	tokenType		type;
+	vString *		string;
+	vString *		scope;
+	unsigned long	lineNumber;
+	MIOPos			filePosition;
+	int 			parentKind; /* -1 if none */
+} tokenInfo;
+
+
+/* Cache variable to build the tag's scope.  It has no real meaning outside
+ * of initPowerShellEntry()'s scope. */
+static vString *FullScope = NULL;
+
+
+static const char *findValidAccessType (const char *const access)
+{
+	unsigned int i;
+	if (access == ACCESS_UNDEFINED)
+		return ACCESS_UNDEFINED; /* early out to save the for-loop if possible */
+	for (i = 0; i < ARRAY_LENGTH(accessTypes); i++)
+	{
+		if (accessTypes[i] == ACCESS_UNDEFINED)
+			continue;
+		if (strcasecmp (access, accessTypes[i]) == 0)
+			return accessTypes[i];
+		i++;
+	}
+	return ACCESS_UNDEFINED;
+}
+
+static void initPowerShellEntry (tagEntryInfo *const e, const tokenInfo *const token,
+								 const powerShellKind kind, const char *const access)
+{
+	int parentKind = -1;
+
+	vStringClear (FullScope);
+
+	initTagEntry (e, vStringValue (token->string));
+
+	e->lineNumber	= token->lineNumber;
+	e->filePosition	= token->filePosition;
+	e->kindName		= PowerShellKinds[kind].name;
+	e->kind			= (char) PowerShellKinds[kind].letter;
+
+	if (access != NULL)
+		e->extensionFields.access = access;
+	if (vStringLength (token->scope) > 0)
+	{
+		parentKind = token->parentKind;
+		if (vStringLength (FullScope) > 0)
+			vStringCatS (FullScope, SCOPE_SEPARATOR);
```

Comment:
Hum, is it me or `FullScope` is actually useless here?  You clear it entering the function, and only add to it once.  In the PHP parser it was used to handle the current namespace separately, but here it doesn't seem needed, is it?
Removing  it will also allow to remove the whole caching as the token (and so, it's inner scope) is guaranteed to live at the very least until `makeTagEntry()` is called.